[Atin Sakkeer] iP

crypto-code opened the pull request:
# WhoBot  😎   
                                                             
 > "Your Future is created by what you do today, not tomorrow" -- Anonymous  

Start creating your future with our **Revolutionary Personal Assistant** that makes managing today's tasks simple. 
```python
                     /$$      /$$ /$$                 /$$$$$$$              /$$    
                    | $$  /$ | $$| $$                | $$__  $$            | $$    
                    | $$ /$$$| $$| $$$$$$$   /$$$$$$ | $$  \ $$  /$$$$$$  /$$$$$$  
                    | $$/$$ $$ $$| $$__  $$ /$$__  $$| $$$$$$$  /$$__  $$|_  $$_/  
                    | $$$$_  $$$$| $$  \ $$| $$  \ $$| $$__  $$| $$  \ $$  | $$    
                    | $$$/ \  $$$| $$  | $$| $$  | $$| $$  \ $$| $$  | $$  | $$ /$$
                    | $$/   \  $$| $$  | $$|  $$$$$$/| $$$$$$$/|  $$$$$$/  |  $$$$/
                    |__/     \__/|__/  |__/ \______/ |_______/  \______/    \___/              

                                                                      
         /$$$$$$ /$$$$$$ /$$$$$$ /$$$$$$ /$$$$$$ /$$$$$$ /$$$$$$ /$$$$$$ /$$$$$$ /$$$$$$ /$$$$$$
        |______/|______/|______/|______/|______/|______/|______/|______/|______/|______/|______/
                                     
```

The WhoBot is
- Super **Fast** and **Efficient**
- Really **Friendly**
- Text-Based
- Well-Designed

## Installation
1. Download the JAR file from [here](https://github.com/crypto-code/ip/releases/download/A-Jar/WhoBot.jar) for _FREE_.
2. Run the JAR file in terminal with the following command, `java -jar WhoBot.jar`
3. Enjoy 😄 

## Features

- [x] Manage ToDos
- [x] Manage Deadlines
- [x] Manage Events
- [x] Search Tasks by Keyword or Date
- [ ] GUI (coming soon)

## For Developers

If you are a developer, you are in luck. The WhoBot is an OpenSource Project and is available [here](https://github.com/crypto-code/ip/tree/master).

This is the `main` method
```java
public static void main(String[] args) {
    WhoBot whoBot = new WhoBot();
    whoBot.run();
}
```

The code is commented to make it easy to navigate and modify for your needs. So Have Fun 😃 

```python

                    /$$$$$$                            /$$ /$$$$$$$                     
                   /$$__  $$                          | $$| $$__  $$                    
                  | $$  \__/  /$$$$$$   /$$$$$$   /$$$$$$$| $$  \ $$ /$$   /$$  /$$$$$$ 
                  | $$ /$$$$ /$$__  $$ /$$__  $$ /$$__  $$| $$$$$$$ | $$  | $$ /$$__  $$
                  | $$|_  $$| $$  \ $$| $$  \ $$| $$  | $$| $$__  $$| $$  | $$| $$$$$$$$
                  | $$  \ $$| $$  | $$| $$  | $$| $$  | $$| $$  \ $$| $$  | $$| $$_____/
                  |  $$$$$$/|  $$$$$$/|  $$$$$$/|  $$$$$$$| $$$$$$$/|  $$$$$$$|  $$$$$$$
                   \______/  \______/  \______/  \_______/|_______/  \____  $$ \_______/
                                                                     /$$  | $$          
                                                                    |  $$$$$$/          
                                                                     \______/           
```

Commit:
Add Gradle support

crypto-code committed:
Level 1

crypto-code committed:
Level 2

crypto-code committed:
Level-3

crypto-code committed:
Level 3

crypto-code committed:
Level-4

crypto-code committed:
Level-4

crypto-code committed:
Level-4

crypto-code committed:
Level-5

crypto-code committed:
Level-5

crypto-code committed:
Level-5

crypto-code committed:
Level-5

crypto-code committed:
Level-5

crypto-code committed:
Level-6

crypto-code committed:
Level-6

crypto-code committed:
Level-7

crypto-code committed:
Level-7

crypto-code committed:
Level-8

crypto-code committed:
Level-8

crypto-code committed:
Level-7 Merge

crypto-code committed:
Merge branch 'branch-Level-8'

crypto-code committed:
Seperate into Different Classes

crypto-code committed:
Seperate into Different Classes

crypto-code committed:
Seperate into Different Classes

crypto-code committed:
Divide classes into packages

crypto-code committed:
Divide classes into packages

crypto-code committed:
Divide classes into packages

crypto-code committed:
Add Tests for Classes

crypto-code committed:
Add JavaDoc comments

crypto-code committed:
Tweak the code to comply with Coding Standard

crypto-code committed:
Add a Find Feature

crypto-code committed:
Merge branch 'branch-Level-9'

crypto-code committed:
Merge branch 'branch-A-JavaDoc'

crypto-code committed:
Add JavaDoc comments

crypto-code committed:
Merge branch 'branch-A-CodingStandard'

crypto-code committed:
Tweak the code to comply with a coding standard

crypto-code committed:
Merge remote-tracking branch 'remotes/origin/add-gradle-support' into branch-A-Gradle

crypto-code committed:
Automate project builds using Gradle

crypto-code committed:
Use CheckStyle

Jai2501 commented on src/main/java/whobot/main/WhoBot.java:
```diff
@@ -0,0 +1,74 @@
+import whobot.utils.TaskList;
+
+
+/**
```
Why is the class comment here starting with `/**`, whereas at other places, it is starting with `/***`?

Jai2501 commented on src/main/java/whobot/utils/TaskList.java:
```diff
@@ -0,0 +1,293 @@
+        }
+    }
+
+    /***
+     * Shows Tasks on a specific date
+     * @param command the command with the date specified
+     * @param ui UI to show output to
+     * @throws WhoBotException If there is any error in the command
+     */
```
Can there be an empty line between description and parameter section?
```suggestion
    /***
     * Shows Tasks on a specific date
     *
     * @param command the command with the date specified
     * @param ui UI to show output to
     * @throws WhoBotException If there is any error in the command
     */
```

Jai2501 commented on src/main/java/whobot/main/UI.java:
```diff
@@ -0,0 +1,99 @@
+public class UI {
+
+    /** Colors used to Display Text */
+    public static final String COLOR_RESET = "\u001B[0m";
+    public static final String COLOR_BLUE = "\u001B[34m";
+    public static final String COLOR_CYAN = "\u001B[36m";
+    public static final String COLOR_PURPLE = "\u001B[35m";
+    public static final String COLOR_RED = "\u001B[31m";
```
Perhaps having these in lexicographical order?

Chen-Yiyang replied:
Very good and clear naming for these constants (colour, commands, etc.). Readability greatly enhanced!

Jai2501 commented on src/main/java/whobot/utils/Helper.java:
```diff
@@ -0,0 +1,168 @@
+    /** UI to which output should be sent */
+    private final UI ui;
+
+    /***
```
Shouldn’t there be an empty line between description and parameter section?

Chen-Yiyang commented on src/main/java/whobot/main/WhoBot.java:
```diff
@@ -0,0 +1,74 @@
+     * Runs the WhoBot and Accepts Input from the User
+     */
+    public void run() {
+        ui.greeting();
```
Maybe you can consider restructure your code here so there is less deep nesting?

Chen-Yiyang commented on src/main/java/whobot/utils/Helper.java:
```diff
@@ -0,0 +1,168 @@
+     * Shows the Main Help Page
+     */
+    public void showMainHelp() {
+        String helpString = "These are the list of commands you can give me:\n";
```
Maybe you can consider avoiding using `+=` for constructing the help message as that seems less obvious what the find result would be?

Chen-Yiyang commented on src/main/java/whobot/main/UI.java:
```diff
@@ -0,0 +1,99 @@
+            + "\t\t                                                   \\______/\n";
+
+    /** Enum Type for indicating the type of echo */
+    public enum Type { START, MIDDLE, END, COMPLETE, ERROR };
```
Nice usage of the `Type` enum to make code easily understandable!

Chen-Yiyang commented on src/main/java/whobot/task/Event.java:
```diff
@@ -0,0 +1,122 @@
+    public Event(String task) throws WhoBotException {
+        super(task.split(" /at ")[0]);
+        try {
+            this.timing = LocalDateTime.parse(task.split(" /at ")[1], DateTimeFormatter.ofPattern("d/M/yyyy HH:mm"));
```
Maybe you can use some local (temporary) variables here, since the expression does a lot of computation and is quite complicated.

Chen-Yiyang reviewed (commented):
Nice and neat code! Great work!

Leofeng10 commented on src/main/java/whobot/utils/Storage.java:
```diff
@@ -0,0 +1,122 @@
+                            + " /at " + tempEvent.getTiming().format(DateTimeFormatter.ofPattern("d/M/yyyy HH:mm")));
+                } else if (type.equals("D")) {
+                    Deadline tempDeadline = (Deadline) tempTask;
+                    dataWriter.write(type + " | " + tempTask.getStatusIcon().charAt(1) + " | " + tempTask.getTask()
+                            + " /by "
+                            + (tempDeadline.hasTime()
+                                    ? tempDeadline.getDeadline().format(DateTimeFormatter.ofPattern("d/M/yyyy HH:mm"))
+                                    : tempDeadline.getDeadline().format(DateTimeFormatter.ofPattern("d/M/yyyy"))));
```
maybe you could break these codes into multiple parts to make it more readable?

Leofeng10 commented on src/main/java/whobot/utils/Parser.java:
```diff
@@ -0,0 +1,88 @@
+
+        String[] commandList = command.toLowerCase(Locale.ROOT).split(" ");
+        //All commands are taken as case-insensitive
+        if (command.toLowerCase(Locale.ROOT).equals("bye") || command.toLowerCase(Locale.ROOT).equals("goodbye")) {
+            // If input is bye or goodbye, quits program
+            ui.goodbye();
+            storage.saveMemory(taskList.getList());
+            return -1;
+        } else if (command.toLowerCase(Locale.ROOT).equals("list")) {
+            // If input is list, prints list
+            taskList.printList(ui);
+        } else if (commandList[0].equals("help")) {
+            // If input is list, prints list
+            if (commandList.length == 1) {
+                helper.showMainHelp();
+            } else {
+                helper.showCommandHelp(commandList[1]);
+            }
+        } else if (commandList.length == 2 && commandList[0].equals("done")) {
+            //If input starts with done, mark the specific item in list as done
+            taskList.markAsDone(commandList[1], ui);
+            storage.saveMemory(taskList.getList());
+        } else if (commandList.length == 2 && commandList[0].equals("undo")) {
+            //If input starts with undo, mark the specific item in list as not done
+            taskList.markAsUndone(commandList[1], ui);
+            storage.saveMemory(taskList.getList());
+        } else if (commandList.length == 2 && commandList[0].equals("delete")) {
+            //If input starts with delete, delete the specific item in list
+            taskList.deleteFromList(commandList[1], ui);
+            storage.saveMemory(taskList.getList());
```
nice work here to give detailed comments

Leofeng10 reviewed (commented):
Great work, and your code is really readable with the comments you provide!

crypto-code committed:
Add GUI

crypto-code committed:
Implement Varargs for Display Buffer

Jai2501 reviewed (commented):
The code is very well written, the variable names are intuitive and enhance the readability of the code. All the coding standard guidelines have been followed. However, a few instances were found where these rules had been violated. LGTM.

crypto-code committed:
Use Assertions

crypto-code committed:
Improve code quality

The code is complex but works well

Code that is complex may not be that readable to other developers plus it may make debugging harder.

Improving Code Quality can improve readability and this will make it easier for others to review the code and also us to debug and make changes more efficently.

Let's clean up the code such that it adheres to the coding standards and also improves code quality. Steps to be taken include simplifying complex expressions, splitting them into multiple statements, avoiding long methods, removing duplicate code and refactoring magic numbers.

By doing these steps, the code quality is vastly improved. By simplifying complex expressions and avoiding long methods, the readability of the code will be improved. By removing duplicates and refactoring magic numbers, the code becomes easier to debug and reuse.

For the coding quality standards followed refer to: https://nus-cs2103-ay2122s1.github.io/website/schedule/week5/topics.html#W5-4

crypto-code committed:
Use Streams

crypto-code committed:
Merge pull request #1 from crypto-code/branch-A-Assertions

Use Assertions

crypto-code committed:
Merge branch 'master' into branch-A-CodeQuality

crypto-code committed:
Merge pull request #3 from crypto-code/branch-A-Streams

Use Streams

crypto-code committed:
Merge pull request #2 from crypto-code/branch-A-CodeQuality

Improve code quality

crypto-code committed:
Set up CI

crypto-code committed:
Set up CI

crypto-code committed:
Set up CI

crypto-code committed:
Improve JUnit Tests

crypto-code committed:
Set up CI

crypto-code committed:
Add Tagging and Anomaly Detection

crypto-code committed:
Improve the GUI

crypto-code committed:
Improve the GUI

crypto-code committed:
Add a User Guide

crypto-code committed:
Add a User Guide

crypto-code committed:
Add files via upload

crypto-code committed:
Update README.md

crypto-code committed:
Set theme jekyll-theme-hacker

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update _config.yml

crypto-code committed:
Update _config.yml

crypto-code committed:
Update README.md

crypto-code committed:
Update _config.yml

crypto-code committed:
Update README.md

crypto-code committed:
Update _config.yml

crypto-code committed:
Add a User Guide

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Add files via upload

crypto-code committed:
Update README.md

crypto-code committed:
Add files via upload

crypto-code committed:
Delete list_GUI.png

crypto-code committed:
Add files via upload

crypto-code committed:
Add files via upload

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Add a User Guide

crypto-code committed:
Add files via upload

crypto-code committed:
Update README.md

crypto-code committed:
Add files via upload

crypto-code committed:
Add files via upload

crypto-code committed:
Add files via upload

crypto-code committed:
Update README.md

crypto-code committed:
Delete delete_GIU.png

crypto-code committed:
Add files via upload

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Update README.md

crypto-code committed:
Add a User Guide

crypto-code committed:
Add a User Guide

crypto-code committed:
Add a User Guide

crypto-code committed:
Release the product

crypto-code committed:
Update README.md

crypto-code committed:
Add files via upload

crypto-code committed:
Add files via upload
